File: 17_subsets/subsets.java
/*
@Copyright:LintCode
@Author:   heiyanbin
@Problem:  http://www.lintcode.com/problem/subsets
@Language: Java
@Datetime: 14-10-18 16:25
*/

class Solution {
    /**
     * @param S: A set of numbers.
     * @return: A list of lists. All valid subsets.
     */
    public ArrayList<ArrayList<Integer>> subsets(ArrayList<Integer> S) {
        // write your code here
        Collections.sort(S);
        DFS(x,S,0);
        return output;

    }
    private void DFS(ArrayList<Integer> x, ArrayList<Integer> S, int start)
    {
        output.add((ArrayList<Integer>)x.clone());
        for(int i=start;i<S.size();i++)
        {
            x.add(S.get(i));
            DFS(x,S,i+1);
            x.remove(x.size()-1);
        }
    }
    private ArrayList<ArrayList<Integer>> output = new ArrayList<ArrayList<Integer>>();
    private ArrayList<Integer> x = new ArrayList<Integer>();
}

File: 17_subsets/subsets.cpp
/*
@Copyright:LintCode
@Author:   heiyanbin
@Problem:  http://www.lintcode.com/problem/subsets
@Language: C++
@Datetime: 14-10-18 16:12
*/

#include <vector>

using namespace std;

class Solution {
public:
    /**
     * @param S: A set of numbers.
     * @return: A list of lists. All valid subsets.
     */
    vector<vector<int> > subsets(vector<int> &nums) {
    	// write your code here
    	vector<int> x;
    	DFS(x,nums,0);
    	return output;
    }
private:
    vector<vector<int>> output;
    void DFS(vector<int> &x, vector<int> &nums,int start)
    {
        output.push_back(x);
        for(int i=start;i<nums.size();i++)
        {
            x.push_back(nums[i]);
            DFS(x,nums,i+1);
            x.pop_back();
        }
    }
    
};
